In [1]:
from math import sqrt
from scipy import stats

%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

## #1. Attendance Data

Load the attendance.csv file and calculate an attendance percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

In [2]:
attendance = pd.read_csv('attendance.csv')

In [3]:
attendance.head()

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [4]:
#first I am renaming the Unnamed column so that it is easier to call later on.
attendance.rename(columns = {'Unnamed: 0':'student_name'}, inplace = True)
attendance.head()

student_name 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0        Sally          P          T          T          H          P   
1         Jane          A          P          T          T          T   
2        Billy          A          T          A          A          H   
3         John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [5]:
# I am now replacing the variables P-Present= 1 day, A-Absent=0 days, T-Tardy= 0.9 days, H-Half= 0.5 days
attendance = attendance.replace({'P': 1.0, 'A': 0.0, 'T': 0.9, 'H': 0.5})
attendance

student_name  2018-01-01  2018-01-02  2018-01-03  2018-01-04  2018-01-05  \
0        Sally         1.0         0.9         0.9         0.5         1.0   
1         Jane         0.0         1.0         0.9         0.9         0.9   
2        Billy         0.0         0.9         0.0         0.0         0.5   
3         John         1.0         0.9         0.5         1.0         1.0   

   2018-01-06  2018-01-07  2018-01-08  
0         0.0         0.9         0.9  
1         0.9         0.0         0.9  
2         0.9         1.0         0.9  
3         0.9         1.0         1.0

In [6]:
# I am now making a new column in the df that holds the total attendance for the 8 day period
attendance['total_attendance'] = attendance.sum(axis=1)/8
attendance

student_name  2018-01-01  2018-01-02  2018-01-03  2018-01-04  2018-01-05  \
0        Sally         1.0         0.9         0.9         0.5         1.0   
1         Jane         0.0         1.0         0.9         0.9         0.9   
2        Billy         0.0         0.9         0.0         0.0         0.5   
3         John         1.0         0.9         0.5         1.0         1.0   

   2018-01-06  2018-01-07  2018-01-08  total_attendance  
0         0.0         0.9         0.9            0.7625  
1         0.9         0.0         0.9            0.6875  
2         0.9         1.0         0.9            0.5250  
3         0.9         1.0         1.0            0.9125

In [7]:
#Finally, I am making a new df of the total attendance per student
total_attendance= attendance[['student_name', 'total_attendance']]
total_attendance

student_name  total_attendance
0        Sally            0.7625
1         Jane            0.6875
2        Billy            0.5250
3         John            0.9125

## #2. Coffee Levels


<li>Read the coffee_levels.csv file.

In [8]:
coffee=pd.read_csv("coffee_levels.csv")
coffee

hour coffee_carafe  coffee_amount
0      8             x       0.816164
1      9             x       0.451018
2     10             x       0.843279
3     11             x       0.335533
4     12             x       0.898291
5     13             x       0.310711
6     14             x       0.507288
7     15             x       0.215043
8     16             x       0.183891
9     17             x       0.391560
10     8             y       0.189297
11     9             y       0.521502
12    10             y       0.023163
13    11             y       0.235529
14    12             y       0.017009
15    13             y       0.997464
16    14             y       0.058361
17    15             y       0.144644
18    16             y       0.544676
19    17             y       0.594126
20     8             z       0.999264
21     9             z       0.915990
22    10             z       0.144928
23    11             z       0.311495
24    12             z       0.771947
25    13             z       0.398520
26    14             z       0.864464
27    15             z       0.436364
28    16             z       0.280621
29    17             z       0.436677

<li>Transform the data so that each carafe is in it's own column.

In [9]:
coffee_transformed = coffee.pivot_table('coffee_amount', 'hour', 'coffee_carafe')
coffee_transformed

coffee_carafe         x         y         z
hour                                       
8              0.816164  0.189297  0.999264
9              0.451018  0.521502  0.915990
10             0.843279  0.023163  0.144928
11             0.335533  0.235529  0.311495
12             0.898291  0.017009  0.771947
13             0.310711  0.997464  0.398520
14             0.507288  0.058361  0.864464
15             0.215043  0.144644  0.436364
16             0.183891  0.544676  0.280621
17             0.391560  0.594126  0.436677

<li>Is this the best shape for the data?
    <br><br>
<i> I do think this is the best shape for the data because the pivot table shows the amount of coffee left in each carafe in each hour. If each carafe represented a different style of coffee, it would be easy to track which coffee type is preferred and used the most. </i>

## 3. Cake Recipes

Read the cake_recipes.csv data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [10]:
cake=pd.read_csv("cake_recipes.csv")
cake.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

Tidy the data as necessary.

In [11]:
#split combined columns (two variables) in to two different columns
recipe_position = cake['recipe:position'].str.split(':', expand=True)
recipe_position.columns = ['recipe', 'position']
# concat the new columns on to the original df
cake = pd.concat([cake,recipe_position], axis=1)
#remove the old column with the combined information
cake = cake.drop(columns='recipe:position')

In [12]:
#next step is to melt data to make it longer rather than wide
cake = cake.melt(id_vars=['recipe', 'position'])
#renaming the columns to understand each key better
cake.rename(columns = {'variable': 'cooking_temperature', 'value': 'overall_rating' }, inplace = True)
cake

recipe position cooking_temperature  overall_rating
0       a   bottom                 225       61.738655
1       a      top                 225       51.709751
2       b   bottom                 225       57.095320
3       b      top                 225       82.455004
4       c   bottom                 225       96.470207
5       c      top                 225       71.306308
6       d   bottom                 225       52.799753
7       d      top                 225       96.873178
8       a   bottom                 250       53.912627
9       a      top                 250       52.009735
10      b   bottom                 250       61.904369
11      b      top                 250       95.224151
12      c   bottom                 250       52.001358
13      c      top                 250       82.795477
14      d   bottom                 250       58.670419
15      d      top                 250       76.101363
16      a   bottom                 275       74.414730
17      a      top                 275       68.576858
18      b   bottom                 275       61.196980
19      b      top                 275       98.594881
20      c   bottom                 275       92.893227
21      c      top                 275       92.098049
22      d   bottom                 275       51.747686
23      d      top                 275       59.571620
24      a   bottom                 300       98.786784
25      a      top                 300       50.228470
26      b   bottom                 300       99.248541
27      b      top                 300       58.169349
28      c   bottom                 300       65.473084
29      c      top                 300       53.960273
30      d   bottom                 300       56.183110
31      d      top                 300       50.971626

Which recipe, on average, is the best? recipe b

In [13]:
cake.groupby('recipe').mean()

overall_rating
recipe                
a            63.922201
b            76.736074
c            75.874748
d            62.864844

Which oven temperature, on average, produces the best results? 275

In [14]:
cake.groupby('cooking_temperature').mean()

overall_rating
cooking_temperature                
225                       71.306022
250                       66.577437
275                       74.886754
300                       66.627655

Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees

In [23]:
cake.groupby(['recipe', 'position', 'cooking_temperature']).mean().sort_values(by="overall_score", ascending=False)

KeyError: 'overall_score'